devito/ir/clusters/algorithms.py

```diff
@@ -198,8 +199,17 @@ def guard(clusters):
                 else:
                     d = cd.parent
 
+                # If `cd` uses, as condition, an arbitrary SymPy expression, then
+                # we must ensure to nest it inside the last of the Dimensions
+                # appearing in `cd.condition`
+                if cd._factor is not None:
+                    k = d
+                else:
+                    dims = pull_dims(cd.condition)
```

Comment:
Here, before pulling, we can evaluate the relation, cuz it could be reduced

Reply:
No, it is late to optimize here. SHould happen way earlier at construction time of ConditionalDimension